fed/api.py

```diff
@@ -217,13 +235,26 @@ def init(
         ping_others(addresses=addresses, self_party=party, max_retries=3600)
 
 
-def shutdown():
+def shutdown(intended=True):
     """
-    Shutdown a RayFed client.
+    Shutdown a RayFed client. Thread safe.
+
+    Args:
+        intended: (Optional) Whether this is a intended exit. If not
+        a "failure handler" will be triggered.
     """
-    compatible_utils._clear_internal_kv()
-    clear_global_context()
-    logger.info('Shutdowned rayfed.')
+    # Since both main thread and data thread will try to shutdown, this
+    # can avoid shutdown twice, so that the failure handler can be
+    # executed only once.
+
+    if (get_global_context() is not None):
```

Comment:
Is `get_global_context` thread safe?

Reply:
No, it's not. The risk only exists when `intended=False`, because other functions except for the user-defined `failure_handler` can be executed multiple times without side effects; and `failure_handler` only gets called when `intended=False` .

As for the time `intended=False`:
1. The entrance of `shutdown(intended=False)` inside RayFed is thread-safe, other calls in driver codes are not; There indeed exists risks that user calls `shutdown(intended=False)` in driver without acquiring the lock, which will accidentally execute the `failure_handler` twice. What do you think if we only expose `shutdown(intended=True)` as API, and make `shutdown(intended=False)` as private method?

2. As described in "The explanation of the _atomic_shutdown_flag", since `shutdown`(where `get_global_context` called) will be entered by OS interrupt's error handler, any lock inside `shutdown` may get hang if it's already been acquired but not yet released before the interrupt. Therefore, making `get_global_context()` thread-safe is not recommended.

Reply:
What if the global context is set to None by other threads?

Reply:
This SHOULD only be done by calling `shutdown`.
Though there's risk that our developer didn't following the rule by mistake, the `shutdown` is indeed can't using lock to be thread-safe.